crates/amalthea/src/kernel.rs

```diff
@@ -78,418 +55,563 @@ pub enum StreamBehavior {
+    };
+    let message = JupyterMessage::create(message, None, &session);
+
+    message.send(&registration_socket)?;
+
+    // Wait for the handshake reply with a 5 second timeout.
+    // If we don't get a handshake reply, we are going to eventually panic and shut down.
+    if !registration_socket
+        .poll_incoming(5000)
+        .map_err(|err| Error::ZmqError(registration_socket.name.clone(), err))?
+    {
+        return Err(crate::anyhow!(
+            "Timeout while waiting for connection information from registration socket"
+        ));
     }
 
-    /// Starts the output capture thread.
-    fn output_capture_thread(iopub_tx: Sender<IOPubMessage>) -> Result<(), Error> {
-        let output_capture = StreamCapture::new(iopub_tx);
-        output_capture.listen();
-        Ok(())
+    // Read the `HandshakeReply` off the socket and confirm its message type
+    let reply = Message::read_from_socket(&registration_socket).unwrap();
+    let status = match reply {
+        Message::HandshakeReply(reply) => reply.content.status,
+        _ => {
+            return Err(crate::anyhow!(
+                "Unexpected message type received from registration socket: {reply:?}"
+            ));
+        },
+    };
+
+    // Check that the client did indeed connect successfully
+    match status {
+        Status::Ok => Ok(()),
+        Status::Error => Err(crate::anyhow!("Client failed to connect to ports.")),
+    }
+}
+
+fn port_finalize(socket: &Socket, port: u16) -> crate::Result<u16> {
+    if port == 0 {
+        // Server provided the port, extract it from the socket
+        // since we gave zmq a port number of `0` to begin with.
+        return port_from_socket(socket);
+    } else {
+        // Client provided the port, just use that
+        return Ok(port);
     }
 }
+
+pub(crate) fn port_from_socket(socket: &Socket) -> crate::Result<u16> {
+    let name = socket.name.as_str();
+
+    let address = match socket.socket.get_last_endpoint() {
+        Ok(address) => address,
+        Err(err) => {
+            return Err(crate::anyhow!(
+                "Can't access last endpoint of '{name}' socket due to {err:?}"
+            ));
+        },
+    };
+
+    let address = match address {
+        Ok(address) => address,
+        Err(_) => {
+            return Err(crate::anyhow!(
+                "Can't access last endpoint of '{name}' socket."
+            ));
+        },
+    };
+
+    // We've got the full address but we only want the port at the very end
+    let Some(loc) = address.rfind(":") else {
+        return Err(crate::anyhow!(
+            "Failed to find port in the '{name}' socket address."
+        ));
+    };
+
+    let port = &address[(loc + 1)..];
+
+    let port = match port.parse::<u16>() {
+        Ok(port) => port,
+        Err(err) => {
+            return Err(crate::anyhow!(
+                "Can't parse port '{port}' into a `u16` due to {err:?}"
+            ));
+        },
+    };
+
+    Ok(port)
+}
```

Comment:
This was kind of ugly but should work for TCP sockets. The only other option is to do what the Python implementation does when they pick a random port, which is to just choose some random port number and try to `bind()` to it. If it works, great, and they know the port number. If it doesn't they pick another one.

We just use `0` and let the OS pick a port, but its kind of hard to get at the port number.